File: Sources/CodexQuotaCore/QuotaSelector.swift
import Foundation

public enum QuotaSelector {
    public static func snapshot(from event: DecodedRateLimitEvent, sourceFingerprint: String) -> QuotaSnapshot? {
        guard event.limitID == "codex" else { return nil }
        guard let window = [event.primary, event.secondary]
            .compactMap({ $0 })
            .first(where: { $0.windowMinutes == 10_080 })
        else {
            return nil
        }
        guard window.usedPercent.isFinite, (0...100).contains(window.usedPercent), window.resetsAt > 0 else { return nil }

        return .init(
            usedPercent: window.usedPercent,
            windowMinutes: window.windowMinutes,
            resetsAt: Date(timeIntervalSince1970: TimeInterval(window.resetsAt)),
            observedAt: event.observedAt,
            sourceFingerprint: sourceFingerprint
        )
    }
}

File: windows/CodexQuota.Windows/QuotaSelector.cs
namespace CodexQuota.Windows;

public static class QuotaSelector
{
    private const int WeeklyWindowMinutes = 10_080;

    public static QuotaSnapshot? Snapshot(DecodedRateLimitEvent value, string sourceFingerprint)
    {
        if (value.LimitId != "codex")
        {
            return null;
        }

        var window = new[] { value.Primary, value.Secondary }
            .FirstOrDefault(candidate => candidate?.WindowMinutes == WeeklyWindowMinutes);
        if (window is null || !double.IsFinite(window.UsedPercent) ||
            window.UsedPercent is < 0 or > 100 || window.ResetsAt <= 0)
        {
            return null;
        }

        DateTimeOffset resetsAt;
        try
        {
            resetsAt = DateTimeOffset.FromUnixTimeSeconds(window.ResetsAt);
        }
        catch (ArgumentOutOfRangeException)
        {
            return null;
        }

        return new QuotaSnapshot(
            window.UsedPercent,
            window.WindowMinutes,
            resetsAt,
            value.ObservedAt,
            sourceFingerprint);
    }
}
